File: src/types/log.ts
export type JSONValue =
    | string
    | number
    | boolean
    | null
    | JSONValue[]
    | { [key: string]: JSONValue };

export type LogType =
    | "BUSINESS-AUDIT"
    | "SECURITY"
    | "COMPLIANCE" // 法令、ガイドライン、等
    | "INFRA"
    | "SYSTEM"
    | "SLA"
    | "DEBUG";
export type LogLevel = 1 | 2 | 3 | 4 | 5 | 6;

export type LogTag = {
    key: string;
    // value: string; // key = value として扱う
    category: string;
};

export interface Log {
    // 識別子・追跡（基本情報）
    traceId: string;
    spanId?: string;
    parentSpanId?: string;
    actorId?: string; // 実行者（UserID, System, or AgentID）

    type: LogType;
    level: LogLevel;
    timestamp: string; // ISO8601 with high precision
    logicalClock: number; // 分散システムでの順序保証用

    // 発生源（トレーサビリティ責務）
    boundary: string; // 発生箇所情報: サービス名やモジュール名
    serviceId: string; // どのインスタンスか（Configから自動注入）

    // 制御メタデータ（安全性責務）
    origin: "SYSTEM" | "AI_AGENT";
    isCritical: boolean;
    aiContext?: {
        agentId: string;
        taskId: string;
        loopDepth: number; // 無限ループ防止用の深度管理
    };

    // コンテンツ（情報責務）
    message: string; // ログ主メッセージ（PIIマスキング・整合性ハッシュ対象）
    input?: JSONValue;
    traceInfo?: string; // TODO 仮
    triggerAgent: boolean;
    agentBackLog?: AIAgentEventBacklog; // AI実行時のみ付与される詳細レコード
    details?: string; // TODO cooperate AI agent

    // 証跡・整合性（不変性責務）
    tags: LogTag[];
    resourceIds?: string[]; // TODO 影響がある口座などの関連情報
    previousHash?: string; // ハッシュチェーン（前のログのハッシュ）
    hash?: string; // このログ自体のハッシュ
    signature?: string; // デジタル署名（非改ざん証明）
}

export interface AIAgentEventBacklog {
    agentId: string; // "anomaly-detector-v2"
    taskId: string; // "task-uuid-123"
    actionType: string; // TODO "analyze", "alert", "remediate"
    model: string; // "gpt-4o", "llama3-70b"
    inputHash: string; // 入力データのハッシュ
    output?: unknown; // TODO AI出力
    isAsynchronous: boolean;
    generatedAt: string;
    triggeredAt?: string;
    processorInfo: AIAgentProcessorInfo;
    confidence?: number; // 信頼度スコア（0.0-1.0）
    status: "pending" | "success" | "failed"; // 実行状態
    error?: string; // エラー詳細
}

export interface AIAgentProcessorInfo {
    resourceInfo: ResourceInfo;
}

export interface ResourceInfo {
    cpu: AllocatedHardwareResourceInfo;
    memory: AllocatedHardwareResourceInfo;
    outerStorage: AllocatedHardwareResourceInfo;
    serviceInfo: ServiceInfo; // in which service the agent is processed.
}

export interface AllocatedHardwareResourceInfo {
    quantity: number;
    unit: string;
}

export interface ServiceInfo {
    serviceId: string; // "payment-service-v1.2.3"
    instanceId: string; // "pod-789-abc123"
    version: string; // "1.2.3"
    deployment: string; // "prod-ap-northeast-1"
    DIContainerRuntime?: string; // "docker", "di-containerd"
}

File: packages/server/internal/domain/log.go
package domain

import "time"

// LogType ログ種別
type LogType string

const (
	LogTypeBusinessAudit LogType = "BUSINESS-AUDIT"
	LogTypeSecurity      LogType = "SECURITY"
	LogTypeCompliance    LogType = "COMPLIANCE"
	LogTypeInfra         LogType = "INFRA"
	LogTypeSystem        LogType = "SYSTEM"
	LogTypeSLA           LogType = "SLA"
	LogTypeDebug         LogType = "DEBUG"
)

// LogLevel ログレベル (1=最低, 6=最高)
type LogLevel int

const (
	LogLevelTrace    LogLevel = 1
	LogLevelDebug    LogLevel = 2
	LogLevelInfo     LogLevel = 3
	LogLevelWarn     LogLevel = 4
	LogLevelError    LogLevel = 5
	LogLevelCritical LogLevel = 6
)

// Origin ログの発生源
type Origin string

const (
	OriginSystem  Origin = "SYSTEM"
	OriginAIAgent Origin = "AI_AGENT"
)

// LogTag キーカテゴリペア
type LogTag struct {
	Key      string
	Category string
}

// AIContext はAIエージェント実行コンテキスト
type AIContext struct {
	AgentID        string
	TaskID         string
	LoopDepth      int
	Model          string
	Confidence     float64
	ReasoningTrace string
}

// AgentBackLogEntry はAIエージェント実行履歴
type AgentBackLogEntry struct {
	AgentID   string
	Action    string
	Timestamp time.Time
	Result    string
	Status    string // "pending", "success", "failed"
}

// Log ドメインモデル
type Log struct {
	TraceID      string
	SpanID       string
	ParentSpanID string
	ActorID      string
	Type         LogType
	Level        LogLevel
	Timestamp    time.Time
	LogicalClock int64
	Boundary     string
	ServiceID    string
	Origin       Origin
	IsCritical   bool
	Message      string
	Tags         []LogTag
	ResourceIDs  []string
	PreviousHash string
	Hash         string
	Signature    string
	// v2: AI agent integration
	AIContext    *AIContext
	AgentBackLog []AgentBackLogEntry
	Input        string
	TriggerAgent bool
	Details      map[string]string
}

// ValidLogTypes 有効なログ種別一覧
var ValidLogTypes = map[LogType]bool{
	LogTypeBusinessAudit: true,
	LogTypeSecurity:      true,
	LogTypeCompliance:    true,
	LogTypeInfra:         true,
	LogTypeSystem:        true,
	LogTypeSLA:           true,
	LogTypeDebug:         true,
}

// IsValidLogType ログ種別の検証
func IsValidLogType(t LogType) bool {
	return ValidLogTypes[t]
}

// IsValidLogLevel ログレベルの検証
func IsValidLogLevel(l LogLevel) bool {
	return l >= 1 && l <= 6
}

// IsValidOrigin オリジンの検証
func IsValidOrigin(o Origin) bool {
	return o == OriginSystem || o == OriginAIAgent
}

// --- Log ドメインメソッド ---

// IP はタグからIPアドレスを取得する。見つからなければデフォルト値を返す。
func (l Log) IP(defaultVal string) string {
	for _, tag := range l.Tags {
		if tag.Key == "ip" {
			return tag.Category
		}
	}
	return defaultVal
}

// FirstResourceID は最初のリソースIDを返す。なければデフォルト値を返す。
func (l Log) FirstResourceID(defaultVal string) string {
	if len(l.ResourceIDs) > 0 {
		return l.ResourceIDs[0]
	}
	return defaultVal
}

// ActorOrDefault はActorIDを返す。空ならデフォルト値を返す。
func (l Log) ActorOrDefault(defaultVal string) string {
	if l.ActorID != "" {
		return l.ActorID
	}
	return defaultVal
}
